services/CurrentTime/include/ble-service-current-time/CurrentTimeService.h

```diff
@@ -0,0 +1,178 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2020 ARM Limited
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef BLE_CURRENT_TIME_SERVICE_H
+#define BLE_CURRENT_TIME_SERVICE_H
+
+#ifdef BLE_FEATURE_GATT_SERVER
+
+#include "ble/BLE.h"
+#include "ble/Gap.h"
+#include "events/EventQueue.h"
+#include "mbed_rtc_time.h"
+
+#include <time.h>
+
+/**
+ * BLE Current Time Service
+ *
+ * @par purpose
+ *
+ * @par usage
+ *
+ * @note The specification for the current time service can be found here:
+ * https://www.bluetooth.com/specifications/gatt
+ *
+ * @attention The user should not instantiate more than a single current time service service
+ */
+class CurrentTimeService : private ble::Gap::EventHandler {
+public:
+    struct CurrentTime {
+        uint16_t year;
+        uint8_t  month;
+        uint8_t  day;
+        uint8_t  hours;
+        uint8_t  minutes;
+        uint8_t  seconds;
+
+        uint8_t  weekday;
+
+        uint8_t  fractions256;
+
+        uint8_t  adjustReason;
+    };
+
+    struct EventHandler {
+        /**
+         * On current time changed
+         *
+         * This function is called if the current time characteristic is changed by the client
+         *
+         * @attention This is an abstract function and should be overridden by the user.
+         */
+        virtual void on_current_time_written(time_t current_time) {}
+    };
+
+    /**
+     * Set event handler
+     *
+     * @param handler EventHandler object to handle events raised by the current time service
+     */
+    void set_event_handler(EventHandler *handler) {
+
+    }
+
+    /**
+     * Construct and initialize a current time service.
+     *
+     * @param ble
+     * @param event_queue
+     * @param chainable_gap_event_handler
+     */
+    CurrentTimeService(BLE &ble, events::EventQueue &event_queue) :
+            _ble(ble),
+            _event_queue(event_queue),
+            _current_time_char(GattCharacteristic::UUID_CURRENT_TIME_CHAR, &_current_time) {
+        memset(&_current_time, 0, sizeof(_current_time));
+    }
+
+    ble_error_t init() {
+        GattCharacteristic *charTable[] = {&_current_time_char};
```

Comment:
this is not specific to this service, but we must find a better way of creating services, we're allocating memory for a characteristic only to have the gatt server allocating more memory and then we keep that characteristic even though the only thing we care about is the handle the gatt server sets on it when it creates the service

there must be sane way to crate a service without this waste

@pan- does the gatt server depends on the existence of the characteristic in any way after the service is added?

Reply:
That's a good point Paul, I think it is safe to create the characteristic on the stack. What we are really interrested in are the handles assigned by the server. 
I need to double check.

Reply:
Answer it is not. I close but that is something we should consider as an improvement for Q1.